addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/SubResourceIntegrityAttributeScanRule.java

```diff
@@ -98,23 +110,65 @@ public static boolean contains(String tag) {
-                        element ->
-                                newAlert()
-                                        .setRisk(Alert.RISK_MEDIUM)
-                                        .setConfidence(Alert.CONFIDENCE_HIGH)
-                                        .setDescription(getString("desc"))
-                                        .setSolution(getString("soln"))
-                                        .setReference(getString("refs"))
-                                        .setEvidence(element.toString())
-                                        .setCweId(345) // CWE-345: Insufficient Verification of Data
-                                        // Authenticity
-                                        .setWascId(15) // Application Misconfiguration
-                                        .raise());
+                        element -> {
+                            String integrityHash = calculateIntegrityHash(msg, element, tree);
+                            String otherInfo =
+                                    integrityHash.isEmpty()
+                                            ? ""
+                                            : Constant.messages.getString(
+                                                    MESSAGE_PREFIX + "otherinfo", integrityHash);
+                            newAlert()
+                                    .setRisk(Alert.RISK_MEDIUM)
+                                    .setConfidence(Alert.CONFIDENCE_HIGH)
+                                    .setDescription(getString("desc"))
+                                    .setSolution(getString("soln"))
+                                    .setReference(getString("refs"))
+                                    .setEvidence(element.toString())
+                                    .setCweId(345) // CWE-345: Insufficient Verification of Data
+                                    // Authenticity
+                                    .setWascId(15) // Application Misconfiguration
+                                    .setOtherInfo(otherInfo)
+                                    .raise();
+                        });
+    }
+
+    private String calculateIntegrityHash(HttpMessage msg, Element element, SiteMap tree) {
+        String src = element.getAttributeValue("src");
+        String integrityHash = "";
+        if (src != null) {
+            URI newURI = null;
+
+            try {
+                String newuri = msg.getRequestHeader().getURI().toString();
+                newURI = new URI(newuri);
+                newURI = newURI.resolve(src);
+                SiteNode node =
+                        tree.findNode(
+                                new org.apache.commons.httpclient.URI(newURI.toString(), true));
+                if (node != null) {
+                    integrityHash =
+                            DigestUtils.sha384Hex(
+                                    node.getHistoryReference()
+                                            .getHttpMessage()
```

Comment:
How do I do this one? Any reading references will be helpful for me.

Reply:
Use `HttpMessage#isResponseFromTargetHost`.

Reply:
Do you mean `node.getHistoryReference().getHttpMessage().isResponseFromTargetHost()`? Under what circumstances would it not be? Redirect?

Reply:
Forwarded messages that are not successful (e.g. timeout).

Reply:
It seems like isResponseFromTagetHost() is always false in the unit tests, do we have to force it or mock it somehow?

I changed like this:
```diff
diff --git a/addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/SubResourceIntegrityAttributeScanRule.java b/addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/SubResourceIntegrityAttributeScanRule.java
index e5f3aeb..c739786 100644
--- a/addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/SubResourceIntegrityAttributeScanRule.java
+++ b/addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/SubResourceIntegrityAttributeScanRule.java
@@ -135,25 +135,25 @@
         if (src == null) {
             return "";
         }
-        // if (msg.isResponseFromTargetHost()) {
         try {
             URI newURI = new URI(msg.getRequestHeader().getURI().toString());
             newURI = newURI.resolve(src);
-            SiteNode node =
-                    tree.findNode(new org.apache.commons.httpclient.URI(newURI.toString(), true));
-            integrityHash =
-                    Base64.getEncoder()
-                            .encodeToString(
-                                    DigestUtils.sha384Hex(
-                                                    node.getHistoryReference()
-                                                            .getHttpMessage()
-                                                            .getResponseBody()
-                                                            .toString())
-                                            .getBytes());
+            SiteNode node = tree
+                    .findNode(new org.apache.commons.httpclient.URI(newURI.toString(), true));
+            HttpMessage scriptNodeMessage = node.getHistoryReference().getHttpMessage();
+            if (scriptNodeMessage.isResponseFromTargetHost()) {
+                integrityHash = Base64.getEncoder()
+                        .encodeToString(
+                                DigestUtils.sha384Hex(
+                                        node.getHistoryReference()
+                                                .getHttpMessage()
+                                                .getResponseBody()
+                                                .toString())
+                                        .getBytes());
+            }
         } catch (Exception e) {
             logger.debug("Error occured while calculating the hash. Error: {}", e.getMessage(), e);
         }
-        // }
         return integrityHash;
     }
 
```

Now 3 tests are failing :( Since `isResponseFromTargetHost` is always false then the calculated hash is always empty string `""` which is not what the test methods are currently expected.

Reply:
Yes I see that I could have used scriptNodeMessage another time for the actual digest :wink:

Reply:
Right, you have to change the mocked message to return the appropriate value (it can be moved to instance variable to be setup as needed in the corresponding tests).

Reply:
Thanks!